test/common/upstream/load_balancer_fuzz_base.cc

```diff
@@ -0,0 +1,127 @@
+
+#include "test/common/upstream/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+void LoadBalancerFuzzBase::initializeASingleHostSet(uint32_t num_hosts_in_host_set,
+                                                    uint8_t index_of_host_set) {
+  ENVOY_LOG_MISC(trace, "Will attempt to initialize host set {} with {} hosts.", index_of_host_set,
+                 num_hosts_in_host_set);
+  MockHostSet& host_set = *priority_set_.getMockHostSet(index_of_host_set);
+  uint32_t hosts_made = 0;
+  // Cap each host set at 250 hosts for efficiency
+  uint32_t max_num_hosts_in_host_set = 250;
+  // Leave port clause in for future changes
+  while (hosts_made < std::min(num_hosts_in_host_set, max_num_hosts_in_host_set) && port_ < 65535) {
+    host_set.hosts_.push_back(makeTestHost(info_, "tcp://127.0.0.1:" + std::to_string(port_)));
+    ++port_;
+    ++hosts_made;
+  }
+  // TODO: Random call and mod against host_set size for locality logic
+}
+
+// Initializes random and fixed host sets
+void LoadBalancerFuzzBase::initializeLbComponents(
+    const test::common::upstream::LoadBalancerTestCase& input) {
+  random_.initialize(input.seed_for_prng());
+  uint8_t index_of_host_set = 0;
+  for (uint16_t num_hosts_in_host_set : input.num_hosts_in_host_set()) {
+    initializeASingleHostSet(num_hosts_in_host_set, index_of_host_set);
+    index_of_host_set++;
+  }
+  num_host_sets_ = index_of_host_set;
+}
+
+// Updating host sets is shared amongst all the load balancer tests. Since logically, we're just
+// setting the mock priority set to have certain values, and all load balancers interface with host
+// sets and their health statuses, this action maps to all load balancers.
+void LoadBalancerFuzzBase::updateHealthFlagsForAHostSet(uint64_t host_index,
+                                                        uint32_t num_healthy_hosts,
+                                                        uint32_t num_degraded_hosts,
+                                                        uint32_t num_excluded_hosts) {
+  uint8_t index_of_host_set = host_index % num_host_sets_;
+  ENVOY_LOG_MISC(trace, "Updating health flags for host set: {}", index_of_host_set);
+  MockHostSet& host_set = *priority_set_.getMockHostSet(index_of_host_set);
+  // This downcast will not overflow because size is capped by port numbers
+  uint32_t host_set_size = host_set.hosts_.size();
+  host_set.healthy_hosts_.clear();
+  host_set.degraded_hosts_.clear();
+  host_set.excluded_hosts_.clear();
+  // The vector constructed here represents remaining indexes, an index will randomly get chosen
+  // from here every time and removed. Thus, this represents the remaining indexes When this goes to
+  // zero, there will be no more indexes left to place into healthy, degraded, and excluded hosts.
+  ENVOY_LOG_MISC(trace, "Host set size: {}", host_set_size);
+  std::vector<uint8_t> indexVector;
+  indexVector.reserve(host_set_size);
+  for (uint8_t i = 0; i < host_set_size; i++) {
```

Comment:
One idea for subset modeling would be a bit vector with offset. The bit vector would be a protobuf `bytes`, which at the given offset would be interpreted bit-wise. That would be fairly compact and surface to the fuzzer ability to go in and mutate in a fine grained way.

Reply:
Hey, that sounds good, but again I feel like the problem with generating "explicit collection of entropy under the control of the fuzzer" is that say this needs to add 80 of 250 hosts to healthy, how would you know how many bytes to generate in fuzzer? What happens if it updates 80 * 8 times?

Reply:
With a 256 max host size, you could peel off 8 bits for an index selection (barring duplicates). I agree with Harvey that this would make it fairly easy for the fuzzer to modulate subset selection. Given it's ability to "figure" out random byte strings like in H/1 or H/2 frames, it would be easy for it to map a selection byte string with the subset selection when it knows every 8 bits changes a host.

> how would you know how many bytes to generate in fuzzer
You might not need to cap the byte string at an exact size. The ideal setting might be 8 bits * 256 hosts, but obviously there may be duplicates, peeling it off after that is basically giving back the 0 index of your available host list.

Reply:
Instead of returning 0 every time, why don't we just "wrap around" the bytestring like I had it before?